### PR TITLE
ath79: add support for linux-4.19 [done]

### PR DESCRIPTION
**Pull request by 981213**
This PR adds linux 4.19 support for ath79.
tested on Winchannel WB2000 (AR9344+AR9382+AR8035).
modifications to ag71xx are tested on 4.14 with the same router.

edit: @mamarley has bumped 910-unaligned_access_hacks.patch and I've included his work here.
edit: I did some comparison between the two usb phy drivers and I'm not sure about what the missing phy_provider is for. Since the usb and phy subsystem is out of my knowledge, I'd like to keep using the version in 4.14 and leave the upstream one untouched.

### CODE REVIEW

**Comment by ynezz** [comment]
@981213 Thanks, works like a charm! Well, almost :-)

I'm experiencing following on my Bullet M/XW which is ar9342:

```
root@OpenWrt:/# uname -a
Linux OpenWrt 4.19.9 #0 Wed Dec 19 14:45:28 2018 mips GNU/Linux


root@OpenWrt:/# cat /tmp/sysinfo/board_name 
ubnt,bullet-m-xw

root@OpenWrt:/# uci set network.lan.ipaddr='192.168.1.20'
root@OpenWrt:/# uci commit network
root@OpenWrt:/# ifup lan
[  152.165245] br-lan: port 1(eth0) entered disabled state
[  152.190413] device eth0 left promiscuous mode
[  152.194955] br-lan: port 1(eth0) entered disabled state
[  152.209672] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[  152.245522] br-lan: port 1(eth0) entered blocking state
[  152.250929] br-lan: port 1(eth0) entered disabled state
[  152.256598] device eth0 entered promiscuous mode
[  152.278819] br-lan: port 1(eth0) entered blocking state
[  152.284142] br-lan: port 1(eth0) entered forwarding state
[  152.390794] br-lan: port 1(eth0) entered disabled state
[  152.407595] device eth0 left promiscuous mode
[  152.412199] br-lan: port 1(eth0) entered disabled state
[  152.461374] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[  152.547608] br-lan: port 1(eth0) entered blocking state
[  152.553002] br-lan: port 1(eth0) entered disabled state
[  152.558695] device eth0 entered promiscuous mode
[  152.620064] br-lan: port 1(eth0) entered blocking state
[  152.625380] br-lan: port 1(eth0) entered forwarding state
'radio0' is disabled
'radio0' is disabled
root@OpenWrt:/# [  158.578307] ------------[ cut here ]------------
[  158.583033] WARNING: CPU: 0 PID: 0 at net/sched/sch_generic.c:461 dev_watchdog+0x16c/0x280
[  158.591437] NETDEV WATCHDOG: eth0 (ag71xx): transmit queue 0 timed out
[  158.598054] Modules linked in: ath9k ath9k_common pppoe ppp_async iptable_nat ipt_MASQUERADE ath9k_hw ath xt_state xt_nat xt_conntrack xt_REDIRECT xt_FLOWOFFLOAD pppox ppp_generic nf_nat_ipv4 nf_nat nf_flow_table_hw nf_flow_table nf_conntrack_rtcache nf_conntrack mac80211 ipt_REJECT cfg80211 xt_time xt_tcpudp xt_multiport xt_mark xt_mac xt_limit xt_comment xt_TCPMSS xt_LOG slhc nf_reject_ipv4 nf_log_ipv4 nf_defrag_ipv6 nf_defrag_ipv4 iptable_mangle iptable_filter ip_tables crc_ccitt compat ip6t_REJECT nf_reject_ipv6 nf_log_ipv6 nf_log_common ip6table_mangle ip6table_filter ip6_tables x_tables ehci_platform ehci_hcd gpio_button_hotplug usbcore nls_base usb_common
[  158.657742] CPU: 0 PID: 0 Comm: swapper Not tainted 4.19.9 #0
[  158.663589] Stack : 80460000 80466d30 80550000 80502d14 80502528 83809dfc 8052fde4 8052fa07
[  158.672104]         804d2c08 00000000 80683670 000001cd 8053a914 00000001 83809db0 00000000
[  158.680608]         00000000 00000000 806b0000 000030e8 00000000 00000000 00000007 00000000
[  158.689122]         000000b3 7d8a455a 000000b2 00000000 00000000 00000000 804fce28 8033aa30
[  158.697616]         00000009 000001cd 8053a914 8053a8d4 00000000 8027fbb4 00000000 80680000
[  158.706123]         ...
[  158.708619] Call Trace:
[  158.711130] [<8006ae74>] show_stack+0x58/0x100
[  158.715663] [<80080b0c>] __warn+0xe4/0x118
[  158.719840] [<80080b70>] warn_slowpath_fmt+0x30/0x3c
[  158.724880] [<8033aa30>] dev_watchdog+0x16c/0x280
[  158.729676] [<800bb7a0>] call_timer_fn+0x20/0x80
[  158.734366] [<800bba74>] run_timer_softirq+0x174/0x1ec
[  158.739625] [<80436010>] __do_softirq+0xe8/0x2c8
[  158.744337] [<80229b60>] plat_irq_dispatch+0xc0/0x120
[  158.749489] [<80065af8>] handle_int+0x138/0x144
[  158.754089] [<80435e5c>] r4k_wait_irqoff+0x18/0x24
[  158.758991] [<800a2dc0>] do_idle+0x78/0xe4
[  158.763152] [<800a3018>] cpu_startup_entry+0x18/0x20
[  158.768198] [<8054fb88>] start_kernel+0x41c/0x43c
[  158.772993] ---[ end trace f46e55e5f4513e69 ]---
[  158.777673] eth0: tx timeout
```

Don't worry too much about it, I'll look into it later once I've more time, but maybe you would've idea where should I best start looking at.

Here is complete dmesg [ar9342-ubnt-bullet-m-xw-dmesg.txt](https://github.com/openwrt/openwrt/files/2695483/ar9342-ubnt-bullet-m-xw-dmesg.txt)

**Comment by mamarley** [comment]
I tried this out on a Ubiquiti UAP-LR and a UAP-AC-PRO and had no issues on either one.

**Comment by 981213** [comment]
@ynezz Is it happen only on 4.19 or both 4.19/4.14? I would suggest a check for PLL values and phy-mode.

**Comment by SibrenVasse** [comment]
Been running this on my TP-Link Archer C7 v2 for the past few days, no issues so far.

**Comment by ynezz** [comment]
@981213 Indeed, I'm able to reproduce it also on 4.14. I'll try ar71xx as next.

**Comment by 981213** [comment]
Rebased and dropped the to-do from my commit message.
I'm not sure about what the phy_provider is for. (It's missing in upstream driver) Since the usb and phy subsystem is out of my knowledge, I'd like to keep using the version in our 4.14 kernel and leave the upstream one untouched.

**Comment by 981213** [comment]
@ynezz According to the machine code "Ubiquiti Bullet M" uses 100Mbps fixed link in ar71xx and in ath79 the link status is polled from PHY. I don't know which one is correct because I have no idea what's the configuration of the actual device.

**Comment by TobleMiner** [comment]
I've tested this PR on ubnt_wa devices and it seem to work just fine. No immediate issues and/or performance degradation. Good work!

**Comment by neheb** [comment]
Just tested this. I can't seem to mount my f2fs drive. I get some OOM kernel panic.

**Comment by neheb** [comment]
Note that my f2fs issues are unrelated to this pull request. This works fine.

**Comment by diizzyy** [comment]
Does USB work in full speed?
https://forum.openwrt.org/t/ath79-target-status/18614

**Comment by neheb** [comment]
On my C7v2, I never saw the not running at full speed message.

**Comment by ynezz** [comment]
I've just merged this to my https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=shortlog;h=refs/heads/staging tree, will try to build and run test it again, then push to master once complete. Thanks!

File: target/linux/ath79/config-4.19 [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_ARCH_USE_QUEUED_RWLOCKS=y
+CONFIG_ARCH_USE_QUEUED_SPINLOCKS=y
+CONFIG_ARCH_WANT_IPC_PARSE_VERSION=y
+CONFIG_ATH79=y
+CONFIG_ATH79_WDT=y
+CONFIG_CC_HAS_SANCOV_TRACE_PC=y
+CONFIG_CC_HAS_STACKPROTECTOR_NONE=y
+CONFIG_CC_IS_GCC=y
+CONFIG_CEVT_R4K=y
+CONFIG_CLANG_VERSION=0
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: Already defined in GENERIC config

File: target/linux/ath79/config-4.19 [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_ARCH_SUSPEND_POSSIBLE=y
+CONFIG_ARCH_USE_BUILTIN_BSWAP=y
+CONFIG_ARCH_USE_QUEUED_RWLOCKS=y
+CONFIG_ARCH_USE_QUEUED_SPINLOCKS=y
+CONFIG_ARCH_WANT_IPC_PARSE_VERSION=y
+CONFIG_ATH79=y
+CONFIG_ATH79_WDT=y
+CONFIG_CC_HAS_SANCOV_TRACE_PC=y
+CONFIG_CC_HAS_STACKPROTECTOR_NONE=y
+CONFIG_CC_IS_GCC=y
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: @ynezz 
Already defined in GENERIC CONFIG

File: target/linux/ath79/config-4.19 [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_ARCH_MMAP_RND_COMPAT_BITS_MAX=15
+CONFIG_ARCH_SUPPORTS_UPROBES=y
+CONFIG_ARCH_SUSPEND_POSSIBLE=y
+CONFIG_ARCH_USE_BUILTIN_BSWAP=y
+CONFIG_ARCH_USE_QUEUED_RWLOCKS=y
+CONFIG_ARCH_USE_QUEUED_SPINLOCKS=y
+CONFIG_ARCH_WANT_IPC_PARSE_VERSION=y
+CONFIG_ATH79=y
+CONFIG_ATH79_WDT=y
+CONFIG_CC_HAS_SANCOV_TRACE_PC=y
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: @ynezz
Already defined in GENERIC config

File: target/linux/ath79/config-4.19 [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_CC_HAS_STACKPROTECTOR_NONE=y
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: @ynezz
Already defined in GENERIC config

File: target/linux/ath79/config-4.19 [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_HARDWARE_WATCHPOINTS=y
+CONFIG_HAS_DMA=y
+CONFIG_HAS_IOMEM=y
+CONFIG_HAS_IOPORT_MAP=y
+CONFIG_HAVE_ARCH_COMPILER_H=y
+CONFIG_HAVE_ARCH_JUMP_LABEL=y
+CONFIG_HAVE_ARCH_KGDB=y
+CONFIG_HAVE_ARCH_SECCOMP_FILTER=y
+CONFIG_HAVE_ARCH_TRACEHOOK=y
+CONFIG_HAVE_CBPF_JIT=y
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: @ynezz 
Is this something we should globally enable? Inconsistenly in the tree

File: target/linux/ath79/config-4.19 [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_HAVE_C_RECORDMCOUNT=y
+CONFIG_HAVE_DEBUG_KMEMLEAK=y
+CONFIG_HAVE_DEBUG_STACKOVERFLOW=y
+CONFIG_HAVE_DMA_CONTIGUOUS=y
+CONFIG_HAVE_DYNAMIC_FTRACE=y
+CONFIG_HAVE_FTRACE_MCOUNT_RECORD=y
+CONFIG_HAVE_FUNCTION_GRAPH_TRACER=y
+CONFIG_HAVE_FUNCTION_TRACER=y
+CONFIG_HAVE_GENERIC_DMA_COHERENT=y
+CONFIG_HAVE_IDE=y
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: Not needed on this platform?

File: target/linux/ath79/config-4.19 [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+CONFIG_OF_NET=y
+CONFIG_PCI_DRIVERS_LEGACY=y
+CONFIG_PERF_USE_VMALLOC=y
+CONFIG_PGTABLE_LEVELS=2
+CONFIG_PHYLIB=y
+# CONFIG_PHY_AR7100_USB is not set
+# CONFIG_PHY_AR7200_USB is not set
+# CONFIG_PHY_ATH79_USB is not set
+CONFIG_PINCTRL=y
+CONFIG_PLUGIN_HOSTCC=""
```
Comment (diizzyy) [inline target/linux/ath79/config-4.19]: @ynezz
Already defined in GENERIC config

**Comment by diizzyy** [comment]
There are probably more duplicates which should be removed.

**Comment by ynezz** [comment]
I've just run `make kernel_oldconfig` and got following output:

```diff
diff --git a/target/linux/ath79/config-4.19 b/target/linux/ath79/config-4.19
index 6545476773..6cdbb968c3 100644
--- a/target/linux/ath79/config-4.19
+++ b/target/linux/ath79/config-4.19
@@ -20,11 +20,7 @@ CONFIG_ARCH_USE_QUEUED_SPINLOCKS=y
 CONFIG_ARCH_WANT_IPC_PARSE_VERSION=y
 CONFIG_ATH79=y
 CONFIG_ATH79_WDT=y
-CONFIG_CC_HAS_SANCOV_TRACE_PC=y
-CONFIG_CC_HAS_STACKPROTECTOR_NONE=y
-CONFIG_CC_IS_GCC=y
 CONFIG_CEVT_R4K=y
-CONFIG_CLANG_VERSION=0
 CONFIG_CLKDEV_LOOKUP=y
 CONFIG_CLONE_BACKWARDS=y
 CONFIG_CMDLINE="rootfstype=squashfs,jffs2"
@@ -58,7 +54,6 @@ CONFIG_DTC=y
 CONFIG_EARLY_PRINTK=y
 CONFIG_ETHERNET_PACKET_MANGLE=y
 CONFIG_FIXED_PHY=y
-CONFIG_GCC_VERSION=70400
 CONFIG_GENERIC_ATOMIC64=y
 CONFIG_GENERIC_CLOCKEVENTS=y
 CONFIG_GENERIC_CMOS_UPDATE=y
@@ -123,7 +118,6 @@ CONFIG_HAVE_OPROFILE=y
 CONFIG_HAVE_PERF_EVENTS=y
 CONFIG_HAVE_REGS_AND_STACK_ACCESS_API=y
 CONFIG_HAVE_RSEQ=y
-CONFIG_HAVE_STACKPROTECTOR=y
 CONFIG_HAVE_SYSCALL_TRACEPOINTS=y
 CONFIG_HAVE_VIRT_CPU_ACCOUNTING_GEN=y
 CONFIG_HW_HAS_PCI=y
@@ -149,6 +143,7 @@ CONFIG_MIGRATION=y
 CONFIG_MIPS=y
 CONFIG_MIPS_ASID_BITS=8
 CONFIG_MIPS_ASID_SHIFT=0
+CONFIG_MIPS_CBPF_JIT=y
 CONFIG_MIPS_CLOCK_VSYSCALL=y
 # CONFIG_MIPS_CMDLINE_BUILTIN_EXTEND is not set
 # CONFIG_MIPS_CMDLINE_DTB_EXTEND is not set
@@ -197,7 +192,6 @@ CONFIG_PHYLIB=y
 # CONFIG_PHY_AR7200_USB is not set
 # CONFIG_PHY_ATH79_USB is not set
 CONFIG_PINCTRL=y
-CONFIG_PLUGIN_HOSTCC=""
 CONFIG_RATIONAL=y
 CONFIG_REGMAP=y
 CONFIG_REGMAP_MMIO=y
```

**Comment by diizzyy** [comment]
That looks better, IDE can most likely be dropped since I can't find any supported devices that uses it. Should reduce the kernel size slightly.

Not sure about the CBPF stuff however

**Comment by ynezz** [comment]
@dizzy I don't want to delay merging of this anymore. It works for quite a few of us, so it's good enough to start testing/fixing, and once it hits git repository, I'm sure, that it gets more attention as not everybody is following GitHub, but a lot of people seems to follow git repository very well, as I've noticed very fast feedback loop(that's very good IMHO) to some of my recent merged PRs.

> That looks better, IDE can most likely be dropped since I can't find any supported devices that uses it. Should reduce the kernel size slightly. Not sure why KVM is enabled either?

I think, that you simply can't [override HAVE_IDE](https://elixir.bootlin.com/linux/v4.19.31/source/arch/mips/Kconfig#L58) and [HAVE_KVM](https://elixir.bootlin.com/linux/v4.19.31/source/arch/mips/Kconfig#L1468) without patching the kernel anyway.

**Comment by neheb** [comment]
I have tested this on my Archer C7v2. Everything works fine.

The only issues that I see is lower iperf speed on ethernet. I believe this is related to the net struct rework that I tried to backport to 4.14 a while ago. Introduced in 4.17.

The f2fs issue mentioned above is unrelated. f2fs is unstable on both big and little endian. I've abandoned it on everything I own (except phones, which is the only place that it gets tested).

**Comment by diizzyy** [comment]
@ynezz 
Hmm, it gets automatically set after "mergering" the config so it's just a matter of preference in the end. Sorry for the noise.

@neheb 
Yeah, f2fs seems quite broken right now :-/

**Pull request closed without merging**